feat: add docs for json schemas (#8705)
withastro/docs: src/content/docs/en/guides/content-collections.mdx

 
See how to convert a basic blog example from `src/pages/posts/` to `src/content/posts` in our [step-by-step tutorial](/en/tutorials/add-content-collections/) that uses the codebase from [the Build a Blog tutorial's finished project](https://github.com/withastro/blog-tutorial-demo).

## Enabling JSON Schema Generation

<p><Since v="4.13.0" /></p>

> If you are working with collections of type `data`, Astro will auto-generate JSON schema files for your editor to get IntelliSense and type-checking. A separate file will be created for each data collection in your project based on your collections defined in `src/content/config.ts` using a library called [zod-to-json-schema](https://github.com/StefanTerdell/zod-to-json-schema#known-issues).

Reviewer: ```suggestion
If you are working with collections of type `data`, Astro will auto-generate JSON schema files for your editor to get IntelliSense and type-checking. A separate file will be created for each data collection in your project based on your collections defined in `src/content/config.ts` using a library called [`zod-to-json-schema`](https://github.com/StefanTerdell/zod-to-json-schema#known-issues).
```

 
See how to convert a basic blog example from `src/pages/posts/` to `src/content/posts` in our [step-by-step tutorial](/en/tutorials/add-content-collections/) that uses the codebase from [the Build a Blog tutorial's finished project](https://github.com/withastro/blog-tutorial-demo).

## Enabling JSON Schema Generation

<p><Since v="4.12.0" /></p>

> If you are working with data collections, you may wish to generate JSON Schema files to get better IntelliSense and type-checking in your editor. Astro provides a flag to auto-generate a JSON Schema for content collections of `type: 'data'` which can be used as the `$schema` value. This feature generates JSON Schema files for each data collection in your project, using a library called [zod-to-json-schema](https://github.com/StefanTerdell/zod-to-json-schema#known-issues)

Reviewer: Normally when we use the word "better", we want to say what we're comparing something to.
Is the point that a JSON schema is better for a data collection than having the schema defined in the `.ts` file like a content collection? Is the *only* purpose for this to improve the editor experience?

Author: The JSON schema is automatically generated, based on the users defined schema.

"better" in terms of DX inside of the editor.

In theory there can be more tools users can add to use that JSON schema for other purposes, but I think 99% of users, use it to improve their DX and editor experience.

Reviewer: Yes, but you still didn't tell me better than WHAT?? I don't know what this is better than, or what the other option is.  :smile:

Author: Better than not having autocompletion and red squiggle lines in the editor 😅

Reviewer: Are you telling me that just having the default schema for data collections currently provides no intellisense? This is what I'm trying to figure out, what currently happens. Data collections are not currently type safe? :smile:

Author: It will not show you the intellisense or any type errors in the editor today. _(It will only show you the type errors in the browser)_

Reviewer: Interesting, OK! I didn't know that was a current limitation of data collections! Thanks!

> To enable this feature, add the flag:

Reviewer: ```suggestion
To enable this feature, add the flag to your `contentCollection` configuration:
```

```astro.config.mjs
import { defineConfig } from 'astro/config';

export default defineConfig({
	contentCollection: {
+		JsonSchema: true
	}
});
```

> This feature requires you to manually reference the schema, which you can either do in each file or in your editor settings:

Reviewer: ```suggestion
This feature requires you to manually set your schema's file path as the value for `$schema` in each data entry file of the collection:
```

> ```src/content/test/entry.json

Reviewer: ``````suggestion
```json title="src/content/authors/armand.json" ins={2}
``````
Now that this is not experimental, can we have a more realistic example here? And format it using Expressive code to highlight the important line?

Author: I'll take a look at the example tomorrow morning _(CET)_

{
  "$schema": "../../../.astro/collections/test.schema.json",
  "test": "test"
}
```

> Each editor has it's own place to configure this, but in VSCode you can use the [`json.schemas` setting](https://code.visualstudio.com/docs/languages/json#_json-schemas-and-settings):

Reviewer: ```suggestion
Alternatively, you can set this value in your editor settings. For example, to set this value in [VSCode's `json.schemas` setting](https://code.visualstudio.com/docs/languages/json#_json-schemas-and-settings), provide the path of files to match and the location of your JSON schema:
```

 
See how to convert a basic blog example from `src/pages/posts/` to `src/content/posts` in our [step-by-step tutorial](/en/tutorials/add-content-collections/) that uses the codebase from [the Build a Blog tutorial's finished project](https://github.com/withastro/blog-tutorial-demo).

## Enabling JSON Schema Generation

> <p><Since v="4.12.0" /></p>

Reviewer: ```suggestion
<p><Since v="4.13.0" /></p>
```
Come back and change this if it's not correct!

> If you are working with collections of type `data`, you can auto-generate JSON schema files for your editor to get IntelliSense and type-checking. A separate file will be created for each data collection in your project based on your collections defined in  `src/content/config.ts` using a library called [zod-to-json-schema](https://github.com/StefanTerdell/zod-to-json-schema#known-issues).

Reviewer: ```suggestion
If you are working with collections of type `data`, you can auto-generate JSON schema files for your editor to get IntelliSense and type-checking. A separate file will be created for each data collection in your project based on your collections defined in `src/content/config.ts` using a library called [zod-to-json-schema](https://github.com/StefanTerdell/zod-to-json-schema#known-issues).
```

> If you are working with data collections, you may wish to generate JSON Schema files to get better IntelliSense and type-checking in your editor. Astro provides a flag to auto-generate a JSON Schema for content collections of `type: 'data'` which can be used as the `$schema` value. This feature generates JSON Schema files for each data collection in your project, using a library called [zod-to-json-schema](https://github.com/StefanTerdell/zod-to-json-schema#known-issues)

Reviewer: ```suggestion
If you are working with collections of type `data`, you can auto-generate JSON schema files for your editor to get IntelliSense and type-checking. A separate file will be created for each data collection in your project based on your collections defined in  `src/content/config.ts` using a library called [zod-to-json-schema](https://github.com/StefanTerdell/zod-to-json-schema#known-issues).
```
Alright, now that I know exactly what is happening, what about something like this to reduce duplicated content and also make it clear that this is a file to be used by your editor (and not somehow a (*replacement* for your actual schema).

The risk you run is it *could* be read as, "Hey, if you're using data collections, you can just automatially generate your schema instead!" and that's in part what I'm trying to avoid here.
